### ISSUE TITLE
Source of lifetime coercion is not reported starting in 1.63

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/rust#99256]
When the lifetime of a value is coerced to a longer one due to the use of an associated function or trait member and the previous lifetime used by other call sites is no longer sufficient, rustc no longer includes any mention of the coercion site (and includes a red herring that forces one to go chasing after the wrong goose). This makes it much, much harder to figure out why code is not working as all the locations rustc mentions are perfectly fine and have nothing to do with the actual cause of the error.

### Code

```rust
use std::path::{Path, PathBuf};

enum PolyStr<'a> {
    Str(&'a str),
}

impl<'a> From<&'static str> for PolyStr<'static> {
    fn from(s: &'static str) -> Self {
        PolyStr::Str(s)
    }
}

#[derive(Default)]
struct Ffcommand<'a> {
    inputs: Vec<&'a Path>,
    output: Option<PolyStr<'a>>,
}

impl<'a> Ffcommand<'a> {
    pub fn run(&self) -> bool {
        true
    }

    pub fn input(mut self, p: &'a Path) -> Self {
        self.inputs.push(p);
        self
    }

    pub fn output<P>(mut self, p: P) -> Self
    where
        P: Into<PolyStr<'a>> + 'a
    {
        self.output = Some(p.into());
        self
    }
}

#[derive(Default)]
struct Ffinfo<'a> {
    cmd: Ffcommand<'a>,
    path: PathBuf,
}

impl Ffinfo<'_> {
    fn foo(&self) -> bool {
        let cmd = Ffcommand::default()
            .input(&self.path)
            // uncomment the next line to see the error
            // .output("pipe:1")
            ;
            
        cmd.run()
    }
}

fn main() {
    let mut info = Ffinfo::default();
    info.path = PathBuf::from("./hello.world");
    info.foo();
}

```

When the commented-out line is added back in, rustc will emit the following error:

```
error[[E0521]](https://doc.rust-lang.org/beta/error-index.html#E0521): borrowed data escapes outside of associated function
  --> src/main.rs:46:19
   |
45 |       fn foo(&self) -> bool {
   |              -----
   |              |
   |              `self` is a reference that is only valid in the associated function body
   |              let's call the lifetime of this reference `'1`
46 |           let cmd = Ffcommand::default()
   |  ___________________^
47 | |             .input(&self.path)
   | |                              ^
   | |                              |
   | |______________________________`self` escapes the associated function body here
   |                                argument requires that `'1` must outlive `'static`

For more information about this error, try `rustc --explain E0521`.
error: could not compile `playground` due to previous error
```

[Try on playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=e27d357e44b18e36503950ba53524c4f).

The error message points to `self` and `.input()` as the locations pertaining to this error, when in fact the error stems entirely from the usage of `.output()`, which coerces the lifetime to `'static` due to the trait impl.

rustc stable (1.62) emits the following error message instead, which is infinitely more helpful:

```
error[[E0759]](https://doc.rust-lang.org/stable/error-index.html#E0759): `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
  --> src/main.rs:47:20
   |
45 |     fn foo(&self) -> bool {
   |            ----- this data with an anonymous lifetime `'_`...
46 |         let cmd = Ffcommand::default()
47 |             .input(&self.path)
   |                    ^^^^^^^^^^ ...is used here...
48 |             .output("pipe:1")
   |              ------ ...and is required to live as long as `'static` here

For more information about this error, try `rustc --explain E0759`.
error: could not compile `playground` due to previous error
```

Notice that here the correct call sites (`.input()` and `.output()`) are mentioned. In particular, the fact that `.output()` is where `'static` is required is clearly mentioned, making it easy to figure out what is going on.

### Version it worked on

It most recently worked on: 1.62.0

### Version with regression

`rustc --version --verbose`: 
```
1.63.0-beta.5 2022-07-10 93ef0cd7fd86d3d05cee
```

**Comment by apiraino** [comment]
bisection seems to lead to commit bb55bd449e65e611da928560d948982d73e50027 #95565 cc @jackh726 

searched nightlies: from nightly-2022-05-01 to nightly-2022-07-27
regressed nightly: nightly-2022-06-08
searched commit range: https://github.com/rust-lang/rust/compare/50b00252aeb77b10db04d65dc9e12ce758def4b5...5435ed6916a59e8d5acba2149316a841c3905cbd
regressed commit: https://github.com/rust-lang/rust/commit/bb55bd449e65e611da928560d948982d73e50027

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.3</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2022-05-01 --script=./script.sh 
```
</details>

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2399256.20Source.20of.20lifetime.20coercion.20is.20not.20reported.20starting.E2.80.A6)).

@rustbot label -I-prioritize +P-high +A-diagnostics

**Comment by mqudsi** [comment]
@apiraino thanks for teaching me about cargo-bisect-rustc 👍

**Comment by estebank** [comment]
This is of course fallout from removing the migrate borrowck.

CC @rust-lang/wg-diagnostics, I believe this should be high priority, but having looked at it in the past it will be non-trivial to get back to the same quality we had before :-/

**Comment by pnkfelix** [comment]
@estebank do you have any ideas on what wg-diagnostics might attempt to do to try to recover the quality here?

**Comment by estebank** [comment]
We need to investigate what E0521 is doing that diverges significantly from E0759. I suspect this won't be trivial to fix :-/
